examples/getting_started/tools_and_agents.ipynb

```diff
@@ -0,0 +1,586 @@
+    "        result. You MUST then call the `extract_content` tool to get the actual content\n",
+    "        of the webpage. It is up to you to determine which search results to parse.\n",
+    "\n",
+    "        Once you have gathered all of the information you need, generate a writeup that\n",
+    "        strikes the right balance between brevity and completeness based on the context of the user's query.\n",
+    "\n",
+    "        MESSAGES: {self.messages}\n",
+    "        USER: {question}\n",
+    "        \"\"\"\n",
+    "    )\n",
+    "    async def _stream(self, question: str) -> openai.OpenAIDynamicConfig:\n",
+    "        return {\n",
+    "            \"tools\": [self.web_search, extract_content],\n",
+    "            \"computed_fields\": {\n",
+    "                \"current_date\": datetime.now().strftime(\"%Y-%m-%d %H:%M:%S\")\n",
+    "            },\n",
+    "        }\n",
+    "\n",
+    "    async def _step(self, question: str):\n",
+    "        response = await self._stream(question)\n",
+    "        tools_and_outputs = []\n",
+    "        async for chunk, tool in response:\n",
+    "            if tool:\n",
+    "                print(f\"using {tool._name()} tool with args: {tool.args}\")\n",
+    "                tools_and_outputs.append((tool, tool.call()))\n",
+    "            else:\n",
+    "                print(chunk.content, end=\"\", flush=True)\n",
+    "        if response.user_message_param:\n",
+    "            self.messages.append(response.user_message_param)\n",
+    "        self.messages.append(response.message_param)\n",
+    "        if tools_and_outputs:\n",
+    "            self.messages += response.tool_message_params(tools_and_outputs)\n",
+    "            await self._step(\"\")\n",
+    "\n",
+    "    async def run(self):\n",
+    "        while True:\n",
+    "            question = input(\"(User): \")\n",
+    "            if question == \"exit\":\n",
+    "                break\n",
+    "            print(f\"(User): {question}\")\n",
+    "            print(\"(Assistant): \", end=\"\", flush=True)\n",
+    "            await self._step(question)\n",
+    "            print()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "This implementation includes:\n",
+    "\n",
+    "1. The `_stream` method, which sets up the LLM call with the necessary tools and computed fields.\n",
+    "2. The `_step` method, which processes the LLM response, handles tool calls, and updates the conversation history.\n",
+    "3. The `run` method, which implements the main interaction loop for the agent.\n",
+    "\n",
+    "The `@openai.call` and `@prompt_template` decorators are used to set up the LLM interaction and define the prompt for the agent. Note that we've explicitly specified the available tools and their usage instructions in the system prompt. This approach has been shown to improve the performance and reliability of tool usage by the LLM.\n",
```

Comment:
nit: you write "This approach has been shown" which leads me to believe there is something like research to share on this point vs. it being something we have found to work in practice, so I would update this instead to say "We have found that this improves the performance..."